src/app/validation/lib/architecture/statistic.ex

```diff
@@ -1,18 +1,39 @@
 defmodule Architecture.Statistic do
   @moduledoc "Behaviour for statistics."
 
+  alias Cloud.Google.Subscription
+
   alias Architecture.LogProvider
   alias Architecture.Resource
   alias Architecture.ResourceSet
   alias Architecture.Timer
 
-  @callback log_providers :: [module]
+  @type t :: module
+  @type message :: any
+
+  @callback providers :: [module] # log providers or other statistics
   @callback resources(ResourceSet.t()) :: ResourceSet.t()
   @callback init(Resource.t()) :: struct
   @callback update(Resource.t(), state) :: state when state: struct
-  @callback handle_log(Resource.t(), state, LogProvider.t(), LogProvider.log()) :: state
+  @callback handle_message(Resource.t(), state, t(), message()) :: state
             when state: struct
 
+  # a statistic can depend on a statistic provider, so cycles are possible
+  @spec check_cycle([module],MapSet.t()) :: :ok
+  def check_cycle(providers,seen) do
+    Enum.each(providers,
```

Comment:
Doesn't this need to fold over the `seen` statistics to be correct? In this case, you are checking if each direct dependency has a cycle, but not that there is no cycle between the entire set of dependencies. At least, I can't seem to convince myself atm that the first fact would imply the second.